### PR TITLE
Implement get_status() RFC

### PR DESCRIPTION
**Pull request by Sami1309**
# Description

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have fixed any merge conflicts

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -205,23 +205,23 @@ stop_postgres:
...
-	pytest client/src/featureform/local_dash_test.py
-	pytest client/tests/redefined_test.py
-	pytest client/tests/local_test.py
-	pytest client/tests/localmode_quickstart_test.py
-	pytest client/tests/register_test.py
-	pytest client/tests/test_spark_provider.py
-	pytest client/tests/test_localmode_include_label_ts.py
-	pytest client/tests/test_localmode_lag_features.py
-	pytest client/tests/wait_resource_test.py
+	# pytest client/tests/status_test.py
```
Comment (sdreyer) [inline Makefile:208]: uncomment

File: client/src/featureform/format.py [hunk 2]
```diff
@@ -16,4 +16,23 @@ def format_rows(format_obj, format_obj_2=None, format_obj_3=None, format_obj_4=N
 
 def format_pg(s=""):
     print(divider)
-    print(s)
\ No newline at end of file
+    print(s)
+
+def format_rows_string(format_obj, format_obj_2=None, format_obj_3=None, format_obj_4=None):
```
Comment (sdreyer) [inline client/src/featureform/format.py:21]: Whats this?

File: client/src/featureform/get.py [hunk 3]
```diff
@@ -7,45 +7,24 @@ def get_user_info(stub, name):
...
-                    f.name, f.variant, "feature")
-            for l in x.labels:
-                format_rows(
-                    l.name, l.variant, "label")
-            for t in x.trainingsets:
-                format_rows(
-                    t.name, t.variant, "training set")
-            format_pg()
-            return x
+            return Entity(
```
Comment (sdreyer) [inline client/src/featureform/get.py:24]: Missing features, labels, and training sets

File: client/src/featureform/get.py [hunk 4]
```diff
@@ -78,134 +57,98 @@ def get_feature_variant_info(stub, name, variant):
...
-            ])
-            format_pg("SOURCE: ")
-            format_rows([("NAME", "VARIANT"), (x.source.name, x.source.variant)])
-            format_pg("TRAINING SETS:")
-            format_rows("NAME", "VARIANT")
-            for t in x.trainingsets:
-                format_rows(t.name, t.variant)
-            format_pg()
-            return x
+            return Feature(
```
Comment (sdreyer) [inline client/src/featureform/get.py:60]: Missing training sets

File: client/src/featureform/get.py [hunk 5]
```diff
@@ -78,134 +57,98 @@ def get_feature_variant_info(stub, name, variant):
...
-            ("STATUS: ", x.status.Status._enum_type.values[x.status.status].name)])
-            format_pg("SOURCE: ")
-            format_rows([("NAME", "VARIANT"), (x.source.name, x.source.variant)])
-            format_pg("TRAINING SETS:")
-            format_rows("NAME", "VARIANT")
-            for t in x.trainingsets:
-                format_rows(t.name, t.variant)
-            format_pg()
-            return x
+            return ff.Label(
```
Comment (sdreyer) [inline client/src/featureform/get.py:79]: Missing training sets

File: client/src/featureform/get.py [hunk 6]
```diff
@@ -78,134 +57,98 @@ def get_feature_variant_info(stub, name, variant):
...
-            for t in x.trainingsets:
-                format_rows(t.name, t.variant)
-            format_pg()
-            return x
+            definition = None
+            is_transformation = None
+            if x.primaryData.table.name:
+                definition = ff.PrimaryData(location=x.primaryData.table.name)
+                is_transformation = False
+            elif x.transformation.SQLTransformation.query:
```
Comment (sdreyer) [inline client/src/featureform/get.py:103]: What about dataframe transformations?

File: client/src/featureform/get.py [hunk 7]
```diff
@@ -78,134 +57,98 @@ def get_feature_variant_info(stub, name, variant):
...
+                definition = ff.SQLTransformation(query=x.transformation.SQLTransformation.query)
+                is_transformation = True
+            return ff.Source(
```
Comment (sdreyer) [inline client/src/featureform/get.py:106]: Missing training sets, features, labels

File: client/src/featureform/get.py [hunk 8]
```diff
@@ -78,134 +57,98 @@ def get_feature_variant_info(stub, name, variant):
...
-            format_pg()
-            return x
+            definition = None
+            is_transformation = None
+            if x.primaryData.table.name:
+                definition = ff.PrimaryData(location=x.primaryData.table.name)
+                is_transformation = False
+            elif x.transformation.SQLTransformation.query:
+                definition = ff.SQLTransformation(query=x.transformation.SQLTransformation.query)
+                is_transformation = True
```
Comment (sdreyer) [inline client/src/featureform/get.py:105]: Whats this for?

File: client/src/featureform/get.py [hunk 9]
```diff
@@ -78,134 +57,98 @@ def get_feature_variant_info(stub, name, variant):
...
-            ("STATUS: ", x.status.Status._enum_type.values[x.status.status].name)])
-            format_pg("LABEL: ")
-            format_rows([("NAME", "VARIANT"), (x.label.name, x.label.variant)])
-            format_pg("FEATURES:")
-            format_rows("NAME", "VARIANT")
-            for f in x.features:
-                format_rows(f.name, f.variant)
-            format_pg()
-            return x
+            return ff.TrainingSet(
```
Comment (sdreyer) [inline client/src/featureform/get.py:123]: Missing provider

File: client/src/featureform/register.py [hunk 10]
```diff
@@ -2572,6 +2572,50 @@ def apply(self):
...
+            "feature": self.get_feature,
+            "label": self.get_label,
+            "source": self.get_source,
+            "trainingset": self.get_training_set,
+            "training-set": self.get_training_set,
+        }
+        
+        resource = None
+        if variant is None:
+            resource = resource_functions[resource_type](name, local=self.local)
```
Comment (sdreyer) [inline client/src/featureform/register.py:2594]: Why would this ever have to wait?

File: client/src/featureform/register.py [hunk 11]
```diff
@@ -2572,6 +2572,50 @@ def apply(self):
...
+        status = resource.get_status()
+        
+        timeout_duration = timedelta(seconds=0)
+        if timeout is not None:
+            timeout_duration = timedelta(seconds=timeout)
+        time_waited = timedelta(seconds = 0)
+        time_started = datetime.now()
+
+        while (status != ResourceStatus.Failed and status != ResourceStatus.Ready) and (timeout is None or time_waited < timeout_duration):
+            if variant is None:
```
Comment (sdreyer) [inline client/src/featureform/register.py:2606]: Make the checks their own function

File: client/src/featureform/register.py [hunk 12]
```diff
@@ -2572,6 +2572,50 @@ def apply(self):
         else:
             state().create_all(self._stub)
 
+    def wait(self, resource_type, name, variant=None, timeout=None):
```
Comment (sdreyer) [inline client/src/featureform/register.py:2575]: Should this return anything?

Comment (sdreyer) [reply client/src/featureform/register.py:2575]: Why not use self.type, self.name, self.variant etc?

File: client/src/featureform/resources.py [hunk 13]
```diff
@@ -853,14 +1012,34 @@ def _create_feature_resource(self, db) -> None:
     def get_status(self):
         return ResourceStatus(self.status)
 
-    def is_ready(self):
-        return self.status == ResourceStatus.READY.value
+    def wait(self, timeout=None):
+        self.wait_function(self.name, self.variant, "feature", timeout)
```
Comment (sdreyer) [inline client/src/featureform/resources.py:1016]: Why do we need to pass these args?

File: client/src/featureform/register.py [hunk 14]
```diff
@@ -2572,6 +2572,50 @@ def apply(self):
...
+        # gets a resource and waits until status is set to ready
+        resource_functions = {
+            "user": self.get_user,
+            "model": self.get_model,
+            "entity": self.get_entity,
+            "provider": self.get_provider
+        }
+
+        resource_variant_functions = {
+            "feature": self.get_feature,
```
Comment (sdreyer) [inline client/src/featureform/register.py:2585]: Isnt this essentially the resource object getting itself? Why not use `self`?

File: client/tests/print_resource_test.py [hunk 15]
```diff
@@ -30,11 +29,9 @@ def training_set_expected_print():
 def test_training_set_print(training_set_object, training_set_expected_print):
        assert training_set_expected_print == f"{training_set_object}"
 
-#Label
-
 @pytest.fixture
 def label_object():
-      return print_outputs["label"]["unit_test"]["object"]
+      return print_outputs["label"]["unit_test"]["resource"]
```
Comment (sdreyer) [inline client/tests/print_resource_test.py:34]: reverted?

File: client/tests/test_outputs.py [hunk 16]
```diff
@@ -1,4 +1,4 @@
-from .resources import Feature, Label, Source, TrainingSet, Entity, Model, Provider, User, ResourceStatus
+from featureform.resources import Feature, Label, Source, TrainingSet, Entity, Model, Provider, User, ResourceStatus, ResourceColumnMapping, PrimaryData, SQLTable, Location, LocalConfig
```
Comment (sdreyer) [inline client/tests/test_outputs.py:1]: Does this need to be changed?

File: pip_update.sh [hunk 17]
```diff
@@ -1,9 +1,30 @@
...
+POSITIONAL_ARGS=()
+
 pip3 uninstall featureform -y
 rm -r client/dist/*
-cd dashboard && npm run build
-cd ../
-mkdir -p client/src/featureform/dashboard/
-cp -r dashboard/out client/src/featureform/dashboard/
+
+# -c or --client parameter has the rebuild only rebuild the client
```
Comment (sdreyer) [inline pip_update.sh:6]: Unclear, should change it to something like --no-dash

File: client/src/featureform/get.py (hunk 4 above)
Comment (Sami1309) [reply client/src/featureform/get.py:63]: added

File: client/src/featureform/get.py (hunk 8 above)
Comment (Sami1309) [reply client/src/featureform/get.py]: fixed. Was a bool field but actually an enum value

File: client/src/featureform/get.py (hunk 6 above)
Comment (Sami1309) [reply client/src/featureform/get.py:108]: added

File: client/src/featureform/get.py (hunk 9 above)
Comment (Sami1309) [reply client/src/featureform/get.py:132]: added

File: client/src/featureform/register.py (hunk 12 above)
Comment (Sami1309) [reply client/src/featureform/register.py]: it does use self.type, variant, etc in the wrapper for each resource in resources.py

File: client/src/featureform/register.py (hunk 14 above)
Comment (Sami1309) [reply client/src/featureform/register.py:2584]: It is. But the ResourceClient has the stub for connection and each function (listed here), and each resource has its data, so this function is passed to each resource, and they provide their data (name, variant, type)

File: client/src/featureform/resources.py (hunk 13 above)
Comment (Sami1309) [reply client/src/featureform/resources.py:1016]: for the function to call the metadata on the resource status.

File: client/tests/print_resource_test.py (hunk 15 above)
Comment (Sami1309) [reply client/tests/print_resource_test.py]: yea, originally it was the wrong name

File: client/src/featureform/register.py (hunk 10 above)
Comment (Sami1309) [reply client/src/featureform/register.py]: removed non variants

File: client/src/featureform/format.py (hunk 2 above)
Comment (sdreyer) [reply client/src/featureform/format.py:21]: Is it possible to make format_rows use this? Take the return from this and print it? Seems like a lot of reused code

File: client/src/featureform/get.py [hunk 18]
```diff
@@ -78,134 +61,105 @@ def get_feature_variant_info(stub, name, variant):
...
-            for t in x.labels:
-                format_rows(t.name, t.variant)
-            format_pg("TRAINING SETS:")
-            format_rows("NAME", "VARIANT")
-            for t in x.trainingsets:
-                format_rows(t.name, t.variant)
-            format_pg()
-            return x
+            is_transformation = None
+            if x.primaryData.table.name:
```
Comment (sdreyer) [inline client/src/featureform/get.py:105]: Should add this to its own function that checks  the transformation type instead of having all the `ifs` out here

File: client/src/featureform/get.py [hunk 19]
```diff
@@ -78,134 +61,105 @@ def get_feature_variant_info(stub, name, variant):
...
+                name=x.name,
+                definition=definition,
+                description=x.description,
+                variant=x.variant,
+                provider=x.provider,
+                owner=x.owner,
+                status=ResourceStatus.from_proto(x.status.status),
+            )
+            if is_transformation is not None:
+                source.is_transformation = is_transformation
```
Comment (sdreyer) [inline client/src/featureform/get.py:124]: This looks like it should be a bool but is getting set as a string.

File: client/src/featureform/register.py [hunk 20]
```diff
@@ -2572,6 +2572,43 @@ def apply(self):
         else:
             state().create_all(self._stub)
 
+    def continue_waiting(status, timeout, time_waited):
```
Comment (sdreyer) [inline client/src/featureform/register.py:2575]: This sounds like an action instead of a check. Should be name something more like `should_wait`

File: client/src/featureform/register.py (hunk 14 above)
Comment (sdreyer) [reply client/src/featureform/register.py:2584]: But cant the resource pass itself in? Why not just have `wait(self, obj, timeout)` and below do `self.wait_function(self, timeout)` ?

File: client/src/featureform/resources.py (hunk 13 above)
Comment (sdreyer) [reply client/src/featureform/resources.py:1024]: See comment above about passing the obj

File: client/tests/get_status_test.py [hunk 21]
```diff
@@ -2,66 +2,77 @@
...
-from .proto import metadata_pb2_grpc, metadata_pb2
+from featureform.proto import metadata_pb2_grpc, metadata_pb2
 
-def resource_with_status(resource, status):
-    resource.status = status
-    return resource
+expected_list = [(metadata_pb2.ResourceStatus.PENDING, ff.ResourceStatus.PENDING), (metadata_pb2.ResourceStatus.READY, ff.ResourceStatus.READY), (metadata_pb2.ResourceStatus.CREATED, ff.ResourceStatus.CREATED), (metadata_pb2.ResourceStatus.FAILED, ff.ResourceStatus.FAILED)]
 
-expected_list = [("PENDING", ff.ResourceStatus.PENDING), ("READY", ff.ResourceStatus.READY), ("CREATED", ff.ResourceStatus.CREATED), ("FAILED", ff.ResourceStatus.FAILED)]
+class StubFeatureVariantGetter:
```
Comment (sdreyer) [inline client/tests/get_status_test.py:13]: Change these to MockFeatureVariantStub, MockTrainingSetVariantStub, etc

File: client/tests/print_resource_test.py [hunk 22]
```diff
@@ -3,108 +3,96 @@
...
 
-from .test_outputs import print_outputs
+from test_outputs import print_outputs
 
 from datetime import timedelta
 
 @pytest.fixture
 def feature_object():
-      return print_outputs["feature"]["unit_test"]["object"]
+    return print_outputs["feature"]["unit_test"]["resource"]
```
Comment (sdreyer) [inline client/tests/print_resource_test.py:12]: Why is this changed back?

File: pip_update.sh [hunk 23]
```diff
@@ -1,9 +1,30 @@
...
+POSITIONAL_ARGS=()
+
 pip3 uninstall featureform -y
 rm -r client/dist/*
-cd dashboard && npm run build
-cd ../
-mkdir -p client/src/featureform/dashboard/
-cp -r dashboard/out client/src/featureform/dashboard/
+
+# -n or --nodash parameter has the rebuild only rebuild the client
```
Comment (sdreyer) [inline pip_update.sh:6]: # -n or --nodash will build the client without the dashboard
